### PR TITLE
[Feature][WIP] JAX Backend

### PR DESCRIPTION
**Pull request by yuanqing-wang**
## Description
Attempt to implement JAX backend.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ x ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/backend/jax/sparse.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+def _addsub(op, x):
+    return -x if op == 'sub' else x
+
+
+def _expand(x, shape):
+    return x.expand(-1, *shape)
+
+def gspmm(gidx, op, reduce_op, X, Y):
+    out, (argX, argY) = _gspmm(gidx, op, reduce_op, X, Y)
```
Comment (yzh119) [inline python/dgl/backend/jax/sparse.py:63]: Not that `_gspmm` and `_gsddmm` do not handle gradients.
In PyTorch backend we use `torch.autograd.Function` to manipulate gradients: https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/sparse.py#L70-L118

Comment (yuanqing-wang) [reply python/dgl/backend/jax/sparse.py:63]: Yes! I was thinking about making sure all the forward passes are working properly before worrying about the gradients. One thing that's a bit tricky is that if we use things like `jax.custom_vjp` we'll have to think about how to cache and reuse stuff from the forward pass like this one: https://github.com/dmlc/dgl/blob/da9f14dff38c9939958ab1126438b7be2c3c39fc/python/dgl/backend/pytorch/sparse.py#L65

**Comment by jermainewang** [comment]
Since this will be a major update, how about we setting up a meeting on how to move forward?

**Comment by yuanqing-wang** [comment]
@jermainewang I'd love to chat as well! When would be a good time to chat?

**Comment by yuanqing-wang** [comment]
Hi @yzh119 just wanted to catch your attention on this again since we had spoken two years ago. 

I've spent some time on this and it looks like this backend is fully functional, passing all unittests. I've also implemented corresponding NN modules. Let me know if you would be interested in taking another look at it!

**Comment by yzh119** [comment]
Thank @yuanqing-wang for the effort, this is huge.

And I love the idea of having a JAX backend, I encourage some full-time DGL team members to take look at this and think about the future of the new backend: @mufeili @Rhett-Ying @BarclayII @jermainewang .

I can help review this PR as well.
